### PR TITLE
Refactor Code Base

### LINKED ISSUE #67: Refactor Codebase
# Description
The codebase requires refactoring prior to proceeding with benchmark, optimization, and recursion sections.

# Steps

- [x] Remove direct index access, e.g. matrix[I], and handle out of index error properly.
- [x] Remove redundant clone() calls to enhance code efficiency.
- [ ] Refactor prover and verifier and update their comments and documentation
- [x] Generate valid witnesses with correct distribution and verify their correctness.
- [x] Return `Transcript` data type instead of `Proof` structure, eliminating unnecessary structs.

### PR DESCRIPTION
**Pull request by omibo**
# Description
This pull request resolves #67  by refactoring the codebase to enhance efficiency, maintainability, and security.

# Changes

- Eliminated all unnecessary `clone()`calls to improve performance.
- Removed the `Proof` struct, moving its elements to the `Transcript` data structure for better organization.
- Removed unnecessary data structures from the `aggregation` module.

### CODE REVIEW

**Comment by omibo** [comment]
# Summary
This comment summarizes the modifications introduced so far for easier review. Up to now, the main refactor is on prover, verifier, and aggregation functionalities.

# Modifications

## Prover

- Removed `Proof` struct and now return the transcript directly (matching the paper). The transcript serves two purposes: absorbing the prover’s messages (to be sent as the proof) and generating Fiat–Shamir randomness.
- Witnesses are now generated in an infinite loop until the witness vector satisfies the l2-norm bound. Previously, this check was inside `prove()` and would panic if the bound failed.
- Vector `t_i` is computed in a separate function instead of inside `prove()`.
- Removed unnecessary `clone()` calls to improve efficiency.
- Switched to the new aggregation API.

## Verifier

- Removed `clone()` calls.
- Switched to the new aggregation API.

## Aggregation

- Reduced code length from 425 lines to 295 lines.
- Added a generic `compute_linear_combination()` function that handles all summation terms across different data types.
- Updated naming to improve readability.

**Comment by omibo** [comment]
# Update
Recent commits resolve issue #65 . All symmetric matrices from the paper are now implemented in a symmetric RqMatrix, enhancing efficiency and soundness.

File: labrador/src/core/garbage_polynomials.rs [hunk 1]
```diff
@@ -1,56 +1,51 @@
...
-                let inner_phi_j_s_i = phi[j].inner_product_poly_vector(&self.witness_vector[i]);
-                h_ij.push(inner_phi_i_s_j + inner_phi_j_s_i);
-            }
-            h_i.push(RqVector::new(h_ij));
+/// Calculate the h_{ij} = <φ_i, s_j> + <φ_j, s_i> garbage polynomials
+/// In the paper, h_{ij} is defined with a factor of 1/2 in front
+/// However, since we're using q = 2^32, division by 2 is problematic in Z_q
+/// So we store h'_{ij} = 2*h_{ij} = <φ_i, s_j> + <φ_j, s_i> directly
+/// Exploits symmetry by only calculating for i ≤ j since h_{ij} = h_{ji}
+pub fn compute_h(witness_vector: &[RqVector], phi: &[RqVector]) -> RqMatrix {
```
[the remark this answers is not in the document]
Comment (omibo) [reply labrador/src/core/garbage_polynomials.rs:28]: Could you clarify this comment? The `test_h_calculation` test currently verifies correctness. Are you suggesting to add zero vectors as an edge case?

File: labrador/src/commitments/ajtai_commitment.rs [hunk 2]
```diff
@@ -44,16 +44,12 @@ pub struct AjtaiScheme {
...
-        witness_bound: Zq,
-        random_matrix: RqMatrix,
-    ) -> Result<Self, ParameterError> {
+    pub fn new(witness_bound: Zq, random_matrix: RqMatrix) -> Result<Self, ParameterError> {
         if witness_bound.is_zero() {
             return Err(ParameterError::InvalidWitnessBounds(witness_bound));
         }
         Self::validate_parameters(
-            beta,
+            witness_bound,
```
[the remark this answers is not in the document]
Comment (omibo) [reply labrador/src/commitments/ajtai_commitment.rs:52]: I fixed it and now only use norm_bound instead of witness_bound or beta.

**Comment by omibo** [comment]
> Very impressive work, Omid! 🔥🔥🔥 Excellent use of lifetimes to make the code more efficient, and great job on all the restructuring. I’ve left a few comments with minor suggestions.
> 
> Overall, this is strong work. When implementing a paper, though, it’s helpful to indicate exactly where each step is coming from. In LaBRADOR, the tables and figures on pages 17–18 make that easy. In many other papers—like Hypernova Folding—that’s not the case.
> 
> It would be nice to briefly point out where this step is discussed in Section 5. No need to do it now, though 🙂

I had started drafting documentation for the prover and verifier. However, the verifier requires some refactoring first. I plan to address this in a separate PR, as this PR already includes so many changes and breaking verifier to smaller functions would make the review process harder. I have added documentation of prover and verifier as a sub-task in #67 .
Please let me know if this makes sense.

File: labrador/src/core/inner_product.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
+use std::{
+    borrow::Borrow,
+    ops::{Add, Mul},
+};
+
+pub fn compute_linear_combination<E, B, C>(elements: &[E], challenges: &[C]) -> B
```
[the remark this answers is not in the document]
Comment (omibo) [reply labrador/src/core/inner_product.rs:50]: I addressed and added documentation for it.

File: labrador/src/core/garbage_polynomials.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (NiDimi) [reply labrador/src/core/garbage_polynomials.rs:28]: Yeah, I am suggesting using the edge case with zeroes. I know we need a lot of edge cases overall, but it was something I wasn thinking while looking at the code

[the remark this answers is not in the document]
Comment (omibo) [reply labrador/src/core/garbage_polynomials.rs:28]: Definitely right. I added zero vector tests for now in the new commit.

File: labrador/src/transcript/sponges/mod.rs [hunk 4]
```diff
@@ -1,6 +1,7 @@
 use crate::ring::{rq::Rq, zq::Zq};
 
 pub trait Sponge {
+    fn default() -> Self;
```
Comment (v0-e) [inline labrador/src/transcript/sponges/mod.rs:4]: Make it implement `Default` instead of adding `default()` inside `Sponge`.
```suggestion
pub trait Sponge: Default {
```

File: labrador/src/ring/rq.rs [hunk 5]
```diff
@@ -49,216 +46,99 @@ impl Rq {
...
+    /// ```compile_fail
+    ///
+    /// let poly = Rq::zero(); // or any constructor you have
+    ///
+    /// // Move happens here …
+    /// let _coeffs = poly.into_coeffs();
+    ///
+    /// // …so using `poly` again must not compile:
+    /// let _still_here = poly; //^ ERROR use of moved value
+    /// ```
```
Comment (v0-e) [inline labrador/src/ring/rq.rs:58]: We can remove this :smile:

**Review by v0-e — APPROVED** [review]
LGTM! 🥇

**Review by NiDimi — APPROVED** [review]
LGTM!
